File: backend/src/main/java/com/example/backend/service/CommissionService.java
package com.example.backend.service;

import com.example.backend.dto.CommissionConfigDTO;
import com.example.backend.dto.CommissionPaiementDTO;
import com.example.backend.dto.LivreurCommissionSummaryDTO;
import com.example.backend.dto.BilanDTO;
import com.example.backend.dto.PageResponse;

import java.math.BigDecimal;
import java.util.List;

public interface CommissionService {

    // ── Commission Config ──────────────────────────────────
    CommissionConfigDTO createConfig(CommissionConfigDTO dto);
    CommissionConfigDTO updateConfig(Long id, CommissionConfigDTO dto);
    CommissionConfigDTO getConfigById(Long id);
    CommissionConfigDTO getActiveConfigByLivreurId(Long livreurId);
    List<CommissionConfigDTO> getConfigsByLivreurId(Long livreurId);
    List<CommissionConfigDTO> getAllActiveConfigs();
    PageResponse<CommissionConfigDTO> searchConfigs(String search, int page, int size);
    void deactivateConfig(Long id);

    // ── Commission Paiement ────────────────────────────────
    CommissionPaiementDTO generateCommission(Long orderId, BigDecimal distanceKm);
    CommissionPaiementDTO getPaiementById(Long id);
    CommissionPaiementDTO getPaiementByOrderId(Long orderId);
    List<CommissionPaiementDTO> getPaiementsByLivreurId(Long livreurId);
    List<CommissionPaiementDTO> getAllPaiements();
    PageResponse<CommissionPaiementDTO> searchPaiements(String search, String status, int page, int size);
    PageResponse<CommissionPaiementDTO> searchPaiementsByLivreur(Long livreurId, String search, String status, int page, int size);
    CommissionPaiementDTO markLivreurPaye(Long paiementId);
    CommissionPaiementDTO markAdminValide(Long paiementId);
    CommissionPaiementDTO unmarkLivreurPaye(Long paiementId);
    CommissionPaiementDTO unmarkAdminValide(Long paiementId);

    // ── Summaries ──────────────────────────────────────────
    LivreurCommissionSummaryDTO getLivreurSummary(Long livreurId);
    List<LivreurCommissionSummaryDTO> getAllLivreurSummaries();

    // ── Recalculate ────────────────────────────────────────
    int recalculateAllDistances();

    // ── Bilan ──────────────────────────────────────────────
    BilanDTO getBilan(Long societeId, Integer annee, Integer mois);
    List<Integer> getAnneesDisponibles(Long societeId);
}

File: front-angular/src/app/services/commission.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { CommissionConfig, CommissionPaiement, LivreurCommissionSummary, BilanDTO } from '../models/commission.model';
import { PageResponse } from '../models/order.model';

@Injectable({
  providedIn: 'root'
})
export class CommissionService {
  private apiUrl = `${environment.apiUrl}/commissions`;

  constructor(private http: HttpClient) {}

  // ── Config ──────────────────────────────────────────────

  createConfig(config: CommissionConfig): Observable<CommissionConfig> {
    return this.http.post<CommissionConfig>(`${this.apiUrl}/configs`, config);
  }

  updateConfig(id: number, config: CommissionConfig): Observable<CommissionConfig> {
    return this.http.put<CommissionConfig>(`${this.apiUrl}/configs/${id}`, config);
  }

  getConfigById(id: number): Observable<CommissionConfig> {
    return this.http.get<CommissionConfig>(`${this.apiUrl}/configs/${id}`);
  }

  getActiveConfigByLivreur(livreurId: number): Observable<CommissionConfig> {
    return this.http.get<CommissionConfig>(`${this.apiUrl}/configs/livreur/${livreurId}`);
  }

  getConfigHistory(livreurId: number): Observable<CommissionConfig[]> {
    return this.http.get<CommissionConfig[]>(`${this.apiUrl}/configs/livreur/${livreurId}/history`);
  }

  getAllActiveConfigs(): Observable<CommissionConfig[]> {
    return this.http.get<CommissionConfig[]>(`${this.apiUrl}/configs/active`);
  }

  searchConfigs(page: number, size: number, search?: string): Observable<PageResponse<CommissionConfig>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (search) params = params.set('search', search);
    return this.http.get<PageResponse<CommissionConfig>>(`${this.apiUrl}/configs/search`, { params });
  }

  deactivateConfig(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/configs/${id}`);
  }

  // ── Paiements ───────────────────────────────────────────

  generateCommission(orderId: number, distanceKm?: number): Observable<CommissionPaiement> {
    let params = new HttpParams();
    if (distanceKm != null) {
      params = params.set('distanceKm', distanceKm.toString());
    }
    return this.http.post<CommissionPaiement>(`${this.apiUrl}/paiements/generate/${orderId}`, null, { params });
  }

  getAllPaiements(): Observable<CommissionPaiement[]> {
    return this.http.get<CommissionPaiement[]>(`${this.apiUrl}/paiements`);
  }

  getPaiementById(id: number): Observable<CommissionPaiement> {
    return this.http.get<CommissionPaiement>(`${this.apiUrl}/paiements/${id}`);
  }

  getPaiementByOrder(orderId: number): Observable<CommissionPaiement> {
    return this.http.get<CommissionPaiement>(`${this.apiUrl}/paiements/order/${orderId}`);
  }

  getPaiementsByLivreur(livreurId: number): Observable<CommissionPaiement[]> {
    return this.http.get<CommissionPaiement[]>(`${this.apiUrl}/paiements/livreur/${livreurId}`);
  }

  searchPaiements(page: number, size: number, search?: string, status?: string): Observable<PageResponse<CommissionPaiement>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (search) params = params.set('search', search);
    if (status) params = params.set('status', status);
    return this.http.get<PageResponse<CommissionPaiement>>(`${this.apiUrl}/paiements/search`, { params });
  }

  searchPaiementsByLivreur(livreurId: number, page: number, size: number, search?: string, status?: string): Observable<PageResponse<CommissionPaiement>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (search) params = params.set('search', search);
    if (status) params = params.set('status', status);
    return this.http.get<PageResponse<CommissionPaiement>>(`${this.apiUrl}/paiements/livreur/${livreurId}/search`, { params });
  }

  markLivreurPaye(paiementId: number): Observable<CommissionPaiement> {
    return this.http.put<CommissionPaiement>(`${this.apiUrl}/paiements/${paiementId}/livreur-paye`, null);
  }

  markAdminValide(paiementId: number): Observable<CommissionPaiement> {
    return this.http.put<CommissionPaiement>(`${this.apiUrl}/paiements/${paiementId}/admin-valide`, null);
  }

  unmarkLivreurPaye(paiementId: number): Observable<CommissionPaiement> {
    return this.http.put<CommissionPaiement>(`${this.apiUrl}/paiements/${paiementId}/unmark-livreur-paye`, null);
  }

  unmarkAdminValide(paiementId: number): Observable<CommissionPaiement> {
    return this.http.put<CommissionPaiement>(`${this.apiUrl}/paiements/${paiementId}/unmark-admin-valide`, null);
  }

  // ── Summaries ───────────────────────────────────────────

  getAllSummaries(): Observable<LivreurCommissionSummary[]> {
    return this.http.get<LivreurCommissionSummary[]>(`${this.apiUrl}/summary`);
  }

  getLivreurSummary(livreurId: number): Observable<LivreurCommissionSummary> {
    return this.http.get<LivreurCommissionSummary>(`${this.apiUrl}/summary/${livreurId}`);
  }

  recalculateAllDistances(): Observable<{ recalculated: number }> {
    return this.http.put<{ recalculated: number }>(`${this.apiUrl}/paiements/recalculate-all`, null);
  }

  // ── Bilan ───────────────────────────────────────────────

  getBilan(annee?: number, mois?: number): Observable<BilanDTO> {
    let params = new HttpParams();
    if (annee != null && annee > 0) params = params.set('annee', annee.toString());
    if (mois != null && mois > 0) params = params.set('mois', mois.toString());
    return this.http.get<BilanDTO>(`${this.apiUrl}/bilan`, { params });
  }

  getAnneesDisponibles(): Observable<number[]> {
    return this.http.get<number[]>(`${this.apiUrl}/bilan/annees`);
  }
}
